pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -779,6 +790,70 @@ func (r *MachineSetSyncReconciler) reconcileMAPItoCAPIMachineSetDeletion(ctx con
 	return true, nil
 }
 
+func (r *MachineSetSyncReconciler) reconcileCAPItoMAPIMachineSetDeletion(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, capiMachineSet *capiv1beta1.MachineSet) (bool, error) {
+	logger := log.FromContext(ctx)
+
+	if capiMachineSet.DeletionTimestamp.IsZero() {
+		if mapiMachineSet == nil || mapiMachineSet.DeletionTimestamp.IsZero() {
+			// Neither CAPI authoritative machine set nor its MAPI non-authoritative mirror
+			// are being deleted, nothing to reconcile for deletion.
+			return false, nil
+		}
+		// The CAPI authoritative machine set is not being deleted, but the MAPI non-authoritative one is
+		// Issue a deletion also to the CAPI authoritative machine set.
+		logger.Info("The non-authoritative Machine API machine set is being deleted, issuing deletion to the corresponding Cluster API machine set")
+
+		if err := r.Client.Delete(ctx, capiMachineSet); err != nil {
+			return false, fmt.Errorf("failed to delete Cluster API machine set: %w", err)
+		}
+
+		// Return true to force a requeue, to allow the deletion propagation.
+		return true, nil
+	}
+
+	if mapiMachineSet == nil {
+		logger.Info("Machine API machine set does not exist, removing corresponding Cluster API machine set sync finalizer")
+		// We don't have  a mapi machine set to clean up. Just let the CAPI operators
+		// function as normal, and remove the CAPI sync finalizer.
+		if _, err := util.RemoveFinalizer(ctx, r.Client, capiMachineSet, machinesync.SyncFinalizer); err != nil {
+			return true, fmt.Errorf("failed to remove finalizer from Cluster API machine set: %w", err)
+		}
+
+		return true, nil
+	}
+
+	if mapiMachineSet.DeletionTimestamp.IsZero() {
+		logger.Info("Cluster API machine set is being deleted, issuing deletion to corresponding Machine API machine set")
+
+		if err := r.Client.Delete(ctx, mapiMachineSet); err != nil {
+			return true, fmt.Errorf("failed delete Machine API machine set: %w", err)
+		}
+
+		return true, nil
+	}
+
+	// TODO: I don't think we run CAPI  controllers, so this won't be present in unit testing.
```

Comment:
Is it ok that we don't test this in unit tests? 

We could run a goroutine somewhere that spots CAPI MachineSets, adds the finalizer and once it sees a deletion timestamp removes it some seconds later?

This doesn't test how our finalizer may interact with the CAPI one though.. (if at all)

Reply:
The test suite should be able to handle the other finalizers yes, I'm sure we've mocked things like this before

I don't think you even need a controller to do it, you can just structure it as part of a specific test case "when the CAPI finalizer is removed" as perhaps a BeforeEach for some further expectations?